pennylane_lightning_gpu/lightning_gpu.py

```diff
@@ -340,18 +351,31 @@ def adjoint_jacobian(self, tape, starting_state=None, use_device_state=False, **
             # whether there must be only one state preparation...
             tp_shift = [i - 1 for i in tp_shift]
 
-        if self._dp.getTotalDevices() > 1 and self._batch_obs:
-            if any(isinstance(ob, _H_dtype(self.C_DTYPE)) for ob in obs_serialized):
-                raise NotImplementedError(
-                    "Hamiltonian observables are not currently supported in multi-GPU adjoint batching work-loads. Please set `batch_obs=False`."
-                )
-
-            jac = adj.adjoint_jacobian_batched(
-                self._gpu_state,
-                obs_serialized,
-                ops_serialized,
-                tp_shift,
+        """
+        This path enables controlled batching over the requested observables, be they explicit, or part of a Hamiltonian.
+        The traditional path will assume there exists enough free memory to preallocate all arrays and run through each observable iteratively.
+        However, for larger system, this becomes impossible, and we hit memory issues very quickly. the batching support here enables several functionalities:
+        - Pre-allocate memory for all observables on the primary GPU (`batch_obs=False`, default behaviour): This is the simplest path, and works best for few observables, and moderate qubit sizes. All memory is preallocated for each observable, and run through iteratively on a single GPU.
+        - Evenly distribute the observables over all available GPUs (`batch_obs=True`): This will evenly split the data into ceil(num_obs/num_gpus) chunks, and allocate enough space on each GPU up-front before running through them concurrently. This relies on C++ threads to handle the orchestration.
+        - Allocate at most `n` observables per GPU (`batch_obs=n`): Providing an integer value restricts each available GPU to at most `n` copies of the statevector, and hence `n` given observables for a given batch. This will iterate over the data in chnuks of size `n*num_gpus`.
+        """
+        if self._batch_obs:
+            num_obs = len(obs_serialized)
+            batch_size = (
+                num_obs
+                if isinstance(self._batch_obs, bool)
```

Comment:
So the batch size is a multiple of `self._dp.getTotalDevices()`? If `self._batch_obs = False`, how does it work? Also, maybe better to understand the intent if we explicitly use an additional variable e.g. `obs_per_gpu`?

Reply:
Yea, this was a tricky one to try to support.

If the user doesn't set batching, or just assumes the default `self._batch_obs = False`, the same behaviour as Lightning occurs: all data is preallocated in the adjoint pipeline, and iterated over in sequence. For smaller problems, this usually runs fine, and has the least overhead.

If a user requests batching by setting the value to true, it evenly splits the observables over all available GPUs, with each batch behaviour the same as the above point.

If an integer value , `n`, is given instead, it limits the number of repeated batches to `n` per GPU, rather than `n` overall. If you have access to multiple GPUs, and have prior knowledge of the system size this will be the most effective overall, as it avoids hitting the memory wall.

While I agree `obs_per_gpu` does make more sense, I did not want to add an additional variable here, as most of this will be deprecated to favour new API designs later.Right now this preserves existing behaviour, with a quick option to set a new value on the fly if needed.